add default.nix which can produce ps and pdf files

apoelstra opened the pull request:
Will also patch the git commit into the "revision" variable which appears in the ps output, if told to build a git repo rather than the current source tree.

apoelstra committed:
Homogenize page source code

Pretty-much just adds and removes some newlines in preparation for
nix-generating the main PostScript file's page numbering.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,218 @@
+let
+  src =
+    if isNull ref
+    then ./.
```
Probably want some sort of filter source here.  I find that `./.` gets really polluted, especially with `result` symlinks from builds.

apoelstra replied:
Sure -- I suspect we actually only want the .ps and .inc files, so this should be an easy filter.

apoelstra replied:
Added a commit which filters the source.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,218 @@
+      '';
+    };
+in
+stdenv.mkDerivation {
```
Why not just put renderBooklet as the derivation result?  Wouldn't that make it easier to check the result into git?  We can do `nix-build -o SSS32.ps` and then commit SSS32.ps. (I haven't actually tried this).

apoelstra replied:
Because I eventually want to produce more than one output -- the full booklet, but also postscript files with individual worksheets, 256-bit vesrions of the worksheets, etc. And maybe a illustrated version alongside the regular version.

apoelstra replied:
On further reflection, and having learned Nix better, let me explore making this derivation have multiple outputs, where the default is the fully-assembled SSS32.ps. I'm unsure if this is the right way to go still, because there are e.g. flags we want to be able to set about PDF vs PS, color vs non-color etc, and it might still be the cleanest thing to output a directory with a bunch of different files.

roconnor-blockstream replied:
My experiments with git and symlinks was a failure.  While we could support multiple outputs, I do not think that it worth worrying about.  At least not at this time.

apoelstra committed:
nix: generate postscript file from include files

Generates the committed postscript and runs diff to make sure that the
reconstruction was faithful. Currently this simply reproduces the
existing code, though it autogenerates the numbers to make it easier
to rearrange things and/or add blank pages.

Later PRs will abstract out the setup code, improve best practices (e.g.
wrapping every page in "10 dict ... end") and add additional targets
which produce e.g. individual worksheets in their own postscript.

We will also use the Nix mechanism to add illustrations, rather than
using PHP.

apoelstra commented:
Force pushed to a .nix which more closely matches the file I'm actually working with to add the full booklet contents, color, etc.

You can see how the dependency management is going to work, once we break up the setup code into more than a single file, and we output the full booklet, but it's hopefully clearer how we'd output other documents too.

roconnor-blockstream commented:
99463184ca456853d5624e3830d439fd58ddd933 seems to add an extra blank line on line 2060 for no good reason.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,247 @@
+        exit 1
+      fi
+    }
+  '' + lib.optionalString doPageChecks toString (
```
Should this all be in a standard `checkPhase` gated by a standard `doCheck` flag?

apoelstra replied:
Ah, yes, it should be.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,222 @@
+  shortId = lib.optionalString (! isNull ref) ("-" + builtins.substring 0 8 src.rev);
+
+
+  setup = rec {
```
probably best to remove `rec`.

apoelstra replied:
I can remove it for now, but it will be used -- this `setup` set will contain all the various components of the postscript "setup" section. Each will have a `dependencies` array which refers to other setup components.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,222 @@
+          ${lib.optionalString (pageData ? drawPageContent) "end\n"}pgsave restore
+          showpage
+        '';
+        nextCtIdx = content.nextCtIdx + (if pageData ? drawPageContent then 1 else 0);
```
Should `CtIdx` just be `PgIdx` - 1, or is that just a coincidence?

apoelstra replied:
It's just a coincidence, but let me check my current development branch to see if we want this calculation at all. It originates with the PostScript originally having a giant array of page content in the setup section, but now that we're assembling pages in Nix, I'm going to move the content into their respective pages rather than indexing into an array.

Regarding the page numbering logic, it's going to wind up being a little ugly, since we need to support

* Pages that get numbers
* Pages that are numbered, but the number isn't displayed because it'd interfere with the content (at least one volvelle page are like this)
* Pages that aren't numbered at all

I'd also like to support some preamble pages having numbers like `i`, `ii`, `iii`, etc.

Anyway for now maybe I just want to rip out the page numbering logic entirely since we need to iterate on it.

apoelstra commented:
@roconnor-blockstream I'll check on the "no good reason" blank line while in flight today, but I suspect that it's there because Nix multiline strings always end in a newline (or is it that they start with a newline? I forget), so if we don't have these newlines in the target output, it forces you to cram lots of stuff onto single lines.

apoelstra commented:
@roconnor-blockstream pushed new version which

* Renames `nextCtIdx` to `nextFooterIdx` to match my local "complete" branch and to reflect that this value indicates the page number to be drawn in the footer, not an index into any content
* Removes the extraneous blank line :)
* Splits the page tests and diff into a `checkPhase` section and added a `doCheck` flag
* Minor refactorings

apoelstra commented on default.nix:
```diff
@@ -0,0 +1,251 @@
+          %*
+          %****************************************************************
+        '' + ''
+          %%Page: ${toString content.nextPgIdx} ${toString content.nextPgIdx}
```
@roconnor-blockstream yep, I'll address that in a future PR. Unfortunately `ps2pdf` loses this information (or maybe PDF doesn't support it) so it's of somewhat limited value.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,221 @@
+
+    mkdir "$out"
+    cd "$out"
+    cp ${renderBooklet fullBooklet} SSS32.ps
```
Should this be a symlink instead?

apoelstra replied:
Hmm, yeah, I guess it could be. It didn't occur to me.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,221 @@
+    cp ${renderBooklet fullBooklet} SSS32.ps
+
+    ${lib.optionalString doOutputDiff "diff -C 5 ${src}/SSS32.ps SSS32.ps"}
+    sed -i 's/(revision \(.*\))/(revision \1${shortId})/' ./SSS32.ps
```
This probably needs to be more robust, but it is fine for now.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,221 @@
+stdenv.mkDerivation {
+  name = "codex32${shortId}";
+
+  buildInputs = if doPdfGeneration then [ ghostscript ] else [ ];
```
I think this ought to be `nativeBuildInputs`.

apoelstra replied:
Ok, will change. I really don't understand the distinction.

roconnor-blockstream replied:
My vague understanding is that `buildInputs` is historic, and means build and runtime dependencies.  While `nativeBuildInputs` is just buildtime dependencies.  I'm not sure what the actual effect of this difference is.

roconnor-blockstream commented on default.nix:
```diff
@@ -199,23 +204,50 @@ let
   buildPhase = ''
     set -e
 
+    ghostscriptTest() {
```
This would make more sense inside `checkPhase`.

apoelstra commented:
Rebased to address commetns.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,251 @@
+    # Copy output Postscript into place
+    mkdir "$out"
+    cd "$out"
+    ln -s "$FULL_BW" SSS32.ps
```
You changed this from `cp` to `ln -s`, but below you patch the file with `sed`.  Does this even work?  the content of `FULL_BW` should be an immutable nix-store file, so I don't think you can patch it.

roconnor-blockstream replied:
Should renderBooklet take a revision argument?

apoelstra replied:
No, it does not work. The `sed` line appears to be doing nothing, and since it does nothing, does not even try to write to the file, and returns success.

As for "should renderBooklet just take a `revision` argument?" let me investigate that. Would be more elegant than rendering the book and then patching it after-the-fact, but might make templating a bit weird.

apoelstra replied:
After testing this by setting the `ref` parameter, I believe the existing code *does* work. I am investigating how. It's as though the `ln -s` is doing a copy-on-write copy rather than a softlink. (I can confirm that the target file is modified by `sed` but the original file is not.)

apoelstra replied:
Oh, `sed -i` will replace linked files unless you pass it `--follow-symlinks`.

Linux.

apoelstra commented:
Will undraft when this is ready for review again. Hopefully today or tomorrow.

apoelstra committed:
nix: introduce single-page tests

apoelstra committed:
add Makefile

apoelstra commented:
Restored the old `cp` line rather than using `ln -s` and relying on `sed`'s magic link-replacing behavior.

Added Makefile which figures out the git commit ID and passes it to nix.

apoelstra commented:
>You may be able to diagnose in 30 seconds and it would be greatly appreciated!

Afraid not. I just checked my local nix CI scripts to see if I got wasm-pack working, and the answer is no, I only have wasm-pack code in one place, and it's all commented out with a confusing explanation. Unfortunately wasm-pack is a nightmare to use in an airgapped context, because it downloads random crapand  does not respect lockfiles.

> Also, is that volvelle-website repo even the correct one to be doing this with? I noticed that this repo seems to have more activity.

Yeah, this repo is where the "real" work is happening. The website is more-or-less a demo and an aid for doing workshops. (I think I will do one in October at TABconf so probably you'll see a bunch of activity there in the week(s) leading up to that, and nothing until then.)

But you raise an interesting point that we should have a self-contained "codex32 tool" that people can build and use locally. WASM is nice because you can run it in the browser and easily have nice interactive interfaces, but it's pretty awful in every other respect.

One soluion for you would be to just build the site and nix-package the binary output....that'd fly in Nix land even though it wouldn't fly in Guix. But I definitely understand if you refuse on principle to do that!

>Do you intend to add some sort of `nix build .#codex32-offline-website` magic to this repo that builds the static website?

Probably not, unless we can get away from wasm-pack. (Which honestly shouldn't be too hard ... in theory I just need to run cargo with a wasm32-unknown-unknown target and then write a tiny bit of glue code ... but I don't know how.)

VzxPLnHqr commented:
> ...see if I got wasm-pack working, and the answer is no... Unfortunately wasm-pack is a nightmare

Thanks for your reply and suggestions. It does give me some comfort that it was not just me who was unable to figure out the wasm-pack stuff with nix!

Regarding a self-contained "codex32 tool," agree that would be helpful.

VzxPLnHqr commented:
I opened a [PR to the website repo](https://github.com/apoelstra/volvelle-website/pull/1) with a flake such that `nix build .#codex32-website` will put a static copy of the website in `./results/www`.

BenWestgate commented:
> > ...see if I got wasm-pack working, and the answer is no... Unfortunately wasm-pack is a nightmare
> 
> Thanks for your reply and suggestions. It does give me some comfort that it was not just me who was unable to figure out the wasm-pack stuff with nix!
> 
> Regarding a self-contained "codex32 tool," agree that would be helpful.

@VzxPLnHqr: What features are you expecting a self-contained "codex32 tool" to have? I may have already made one.

VzxPLnHqr commented:
> What features are you expecting a self-contained "codex32 tool" to have? I may have already made one.

@BenWestgate I have not played with it too much yet, so maybe a self-hosted / static version of the website will suffice? (such as as what I added to [nixos-airgapped](https://github.com/VzxPLnHqr/nixos-airgapped) ). But a simple command line interface could be nice too. Is that what you have made already?

BenWestgate commented:
> > What features are you expecting a self-contained "codex32 tool" to have? I may have already made one.
> 
> @BenWestgate I have not played with it too much yet, so maybe a self-hosted / static version of the website will suffice? (such as as what I added to [nixos-airgapped](https://github.com/VzxPLnHqr/nixos-airgapped) ). But a simple command line interface could be nice too. Is that what you have made already?

Yes, https://github.com/BenWestgate/codex32/blob/master/reference/python-codex32/src/codex32.py has functions for all sorts of codex32 needs.

I started working on a Gtk GUI for importing with error correction: https://github.com/BenWestgate/bails-wallet/blob/master/bails-wallet/string_entry.py I think it was meeting most or all of the criteria of https://github.com/BlockstreamResearch/codex32/blob/master/docs/wallets.md when I last tested it.

These were originally for my project https://github.com/BenWestgate/Bails which uses the python library above to support creating codex32 backups and importing them to Bitcoin core. If andrew's bitcoin/bitcoin future PR https://github.com/bitcoin/bitcoin/pull/27351#issuecomment-2059322038 gets merged most of that code except my GUI and error correction can be deleted.

roconnor-blockstream commented:
`Page: 1 1` seems to now have `portraitPage begin` twice.  Presumably this is an error

apoelstra commented:
Oh, good catch. This appears in `title.inc.ps` but shouldn't because the default.nix code now does it.

apoelstra commented:
Fixed.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,251 @@
+    else
+      fetchGit {
+        url = ./.;
+        inherit ref;
```
After reviewing https://nix.dev/manual/nix/2.22/language/builtins#builtins-fetchGit I'm pretty sure you want `rev` instead of `ref`.

apoelstra replied:
Empirically `rev` seems to want a 40-character hash, while `ref` will accept short hashes, branch names, etc.

roconnor-blockstream commented:
> Oh, good catch. This appears in `title.inc.ps` but shouldn't because the default.nix code now does it.

Maybe I'm being too picky, but you patched the wrong commit.  You should have put the `title.inc.ps` patch in `add version to footer; use portraitPage and landscapePage on every page`.  However instead you patched `nix: generate postscript file from include files`.

`nix: generate postscript file from include files` was supposed to be a nop with regards to the `SSSS32.ps` file, but now that files has a diff in it, a diff that gets reversed again in `add version to footer; use portraitPage and landscapePage on every page`.

apoelstra commented:
I don't think you're being too picky. Though I'm surprised I made a mistake like that, given that I used `git blame` to zero in on what I thought was the correct commit, and then ran `nix-build` on every subsequent commit which showed that the SSSS32.ps file was consistent with the nixfile.

I'll fix the history. Sorry about that.

apoelstra committed:
add version to footer; use portraitPage and landscapePage on every page

apoelstra commented:
Ah, I see the issue. I was misled by `git blame` (or perhaps, misled by my weird PR structure where I introduced code intending to later modify it).

Fixed.

roconnor-blockstream commented:
Are you familiar with `gitattributes`'s `export-subst`?

[Bitcoin uses this](https://github.com/bitcoin/bitcoin/blob/master/.gitattributes) to insert a [define `GIT_COMMIT_ID`](https://github.com/bitcoin/bitcoin/blob/master/src/clientversion.cpp#L33) into this file.  This method might make for a better way adding shortids to the (generated) postscript.

apoelstra commented:
No, I am not familiar -- though it looks like it needs `git-archive` to work. I am averse to using git-archive, especially the way Bitcoin Core does, because it interacts with nix in unpredictable ways. (For example in rust-bitcoinconsensus when I try to verify that the vendored Core code matches the relevant Core commit, I have a giant list of whitelisted files which seem to be .gitignored in one case but not in the other, and files which always have diffs that I can't reproduce locally, presumably because they are done by git attributes.)

I can go learn all these things if you want but I feel that my existing way, which behaves predictably and uses explicit code, will be easier to maintain.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,251 @@
+
+  nativeBuildInputs = if doPdfGeneration then [ ghostscript ] else [ ];
+
+  phases = [ "buildPhase" ] ++ lib.optionals doCheck [ "checkPhase" ];
```
Unfortunately the `checkPhase` is never run because `doCheck` needs to be set to `true` in the `stdenv.mkDerivation` in order to run the `checkPhase`

You should change the phases line to:

    phases = [ "buildPhase" "checkPhase" ];
    inherit doCheck;

and let `doCheck` control whether or not the check phase is executed or not.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,251 @@
+    ${lib.optionalString doPdfGeneration "ps2pdf -dPDFSETTINGS=/prepress SSS32.ps"}
+  '';
+
+  checkPhase = toString
```
Unfortunately the check fails when it is actually enabled, at least in part because you try to run `ghostscriptTest` before it is defined.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,251 @@
+
+  checkPhase = toString
+    (map
+      (page: "ghostscriptTest ${checkSinglePage page}")
```
There is another issue with this design.  Because `checkPhase` contains references to the single pages, those single pages must be built, whether or not they will be checked.  In this case I suggest taking the unusual step of conditionally excluding the entire contents of the `checkPhase` value if `doCheck` is false.

Alternatively you could replace `fullBooklet.pages` with `[]` when `doCheck` is false.

roconnor-blockstream commented on default.nix:
```diff
@@ -0,0 +1,251 @@
+  buildPhase = ''
+    set -e
+
+    FULL_BW="${renderBooklet fullBooklet}"
```
The way nix derivations work is that, more or less, every value in the derivation becomes an environment variable.

Therefore, it would make more sense for you to define `FULL_BW` as an entry in the derivation rather that setting it in the build phase.
